### PR TITLE
 Deserialisation fixes and PaymentMethodType update

### PR DESCRIPTION
**Pull request by CaseyODonnell92**

### CODE REVIEW

File: src/main/java/io/electrum/vas/model/An32TokenPayment.java [hunk 1]
```diff
@@ -50,7 +56,9 @@ public String toString() {
       sb.append("class An32TokenPayment {\n");
 
       sb.append("    type: ").append(Utils.toIndentedString(type)).append("\n");
+      sb.append("    name: ").append(Utils.toIndentedString(name)).append("\n");
       sb.append("    token: ").append(Utils.toIndentedString(token)).append("\n");
+      sb.append("    amount: ").append(Utils.toIndentedString(amount)).append("\n");
```
Comment (TimElectrum) [inline src/main/java/io/electrum/vas/model/An32TokenPayment.java]: maybe move this above token so that all the parent fields are together

File: src/main/java/io/electrum/vas/model/BasicAdvice.java [hunk 2]
```diff
@@ -18,7 +18,7 @@
 /**
  * The data required in all value added service requests
  **/
-@ApiModel(description = "The data required in all value added service requests", subTypes = {TenderAdvice.class, BasicReversal.class})
+@ApiModel(description = "The data required in all value added service requests")
```
Comment (TimElectrum) [inline src/main/java/io/electrum/vas/model/BasicAdvice.java]: why did you remove the subtype descriptions? I gather you had a reason after all your looking into inheritance in swagger etc. but I'd like to learn from it - i'm too lazy to look it up myself

Comment (TimElectrum) [inline src/main/java/io/electrum/vas/model/BasicAdvice.java]: also, please make this description better. it's just plan wrong as it is.

File: src/main/java/io/electrum/vas/model/LoyaltyCardPayment.java [hunk 3]
```diff
@@ -50,7 +57,9 @@ public String toString() {
       sb.append("class LoyaltyCardPayment {\n");
 
       sb.append("    type: ").append(Utils.toIndentedString(type)).append("\n");
+      sb.append("    name: ").append(Utils.toIndentedString(name)).append("\n");
       sb.append("    cardNumber: ").append(Utils.toIndentedString(cardNumber)).append("\n");
+      sb.append("    amount: ").append(Utils.toIndentedString(amount)).append("\n");
```
Comment (TimElectrum) [inline src/main/java/io/electrum/vas/model/LoyaltyCardPayment.java]: again, put the amount before the cardNumber field

File: src/main/java/io/electrum/vas/model/PaymentMethod.java [hunk 4]
```diff
@@ -11,10 +13,14 @@
 import java.util.Objects;
 
 @ApiModel(description = "Base model for all payment types", discriminator = "type", subTypes = {An32TokenPayment.class, LoyaltyCardPayment.class})
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "type")
```
Comment (TimElectrum) [inline src/main/java/io/electrum/vas/model/PaymentMethod.java]: thanks for looking into this - it's going to be great having proper inheritance in our swagger

File: src/main/java/io/electrum/vas/model/BasicAdvice.java (hunk 2 above)
Comment (CaseyODonnell92) [reply src/main/java/io/electrum/vas/model/BasicAdvice.java]: In the case of PaymentMethods, we wanted to be able to just have PaymentMethod as a field somewhere in the API definition, and by virtue of that, have all the subtypes also defined explicitly in the generated swagger spec. This currently happens in the case of PaymentMethods. In the case of BasicAdvice, it was not clear that we wanted this behaviour - other service interfaces (as far as I'm aware) explicitly reference the kinds of advice they need (and the fields are still inherited and correctly defined in the swagger spec). Without a discriminator field, it is not entirely clear if deserialisation will be deterministic. So basically, swagger inheritance is not clearly necessary in this case, and it might make de/serialisation more complex.
